### PR TITLE
added results cache

### PR DESCRIPTION
**Pull request by cbizon**
Adds a redis results cache to aragorn.   Probably acts badly if it can't reach such a cache...

Includes an override_cache and timeout_seconds parameters for controlling the cache's use.

Tested against aragorn and robokop.

You can see usage here: https://github.com/ranking-agent/robogallery/blob/master/Integration/ARAGORN_Cache.ipynb

Still needs to have the helm chart updated to include the redis and secrets etc.  I don't have a good sense of how much memory is needed, and I haven't explored the idea of compressing the trapi results in the redis.

### CODE REVIEW

File: src/service_aggregator.py [hunk 1]
```diff
@@ -120,8 +121,9 @@ async def entry(message, guid, coalesce_type, caller) -> (dict, int):
         del message["workflow"]
     else:
         if infer:
+            timeout_seconds = message.get("parameters", {}).get("timeout_seconds", 3 * 60)
```
Comment (kennethmorton) [inline src/service_aggregator.py:124]: Need any other emergency type checking here?

File: src/service_aggregator.py [hunk 2]
```diff
@@ -142,8 +144,24 @@ async def entry(message, guid, coalesce_type, caller) -> (dict, int):
     # Raise a 422 if we find one we don't actually know how to do.
     # We told the world what we can do!
     # Workflow will be a list of the functions, and the parameters if there are any
+
+    results_cache = ResultsCache()
+    override_cache = message.get("parameters",{}).get("override_cache", False)
```
Comment (kennethmorton) [inline src/service_aggregator.py:149]: Type checking? I'm worried about override_cache == "True"

File: src/service_aggregator.py [hunk 3]
```diff
@@ -165,7 +186,7 @@ def is_end_message(message):
     return False
 
 
-async def post_with_callback(host_url, query, guid, params=None):
+async def post_with_callback(host_url, query, guid, params):
```
Comment (kennethmorton) [inline src/service_aggregator.py:189]: Why did this change from keyword arg to required? will this break anything?

File: src/service_aggregator.py [hunk 4]
```diff
@@ -251,7 +272,11 @@ async def collect_callback_responses(guid, num_queries):
 
     # Don't spend any more time than this assembling messages
     #OVERALL_TIMEOUT = timedelta(hours=1)  # 1 hour
-    OVERALL_TIMEOUT = timedelta(minutes=3)  # 3 minutes
+    logger.info(f"{guid}: params: {params}")
+    if "timeout_seconds" in params:
+        OVERALL_TIMEOUT = timedelta(seconds=params["timeout_seconds"])
```
Comment (kennethmorton) [inline src/service_aggregator.py:277]: type check on timeout_seconds.

**Review by kennethmorton — APPROVED** [review]

File: src/service_aggregator.py (hunk 1 above)
Comment (maximusunc) [reply src/service_aggregator.py]: Added

File: src/service_aggregator.py (hunk 2 above)
Comment (maximusunc) [reply src/service_aggregator.py]: Added

File: src/service_aggregator.py (hunk 4 above)
Comment (maximusunc) [reply src/service_aggregator.py]: I followed the flow around and at this point, params is either an empty dict or the timeout_seconds type has been checked already. So I simplified this logic a little.

File: src/service_aggregator.py (hunk 3 above)
Comment (maximusunc) [reply src/service_aggregator.py]: I changed this to an empty dict and removed the POST logic below this. POSTing with params doesn't actually work.

**Pull request merged**